sql/jsonpath: validate JSON column in jsonb_path_exists for inverted index acceleration

ZhouXing19 opened the pull request:
Informs: #154729

Previously, the `jsonb_path_exists` function would attempt inverted index acceleration even when the JSON parameter referenced a column that was not the source column of the inverted index, or when it was a constant JSON value. This resulted in incorrect query plans and potentially wrong results.

This commit adds proper validation to ensure that inverted index acceleration is only used when the JSON parameter is a variable expression that references the actual source column of the inverted index. For all other cases, the query will fall back to a full table scan with post-filtering. 

We also don't allow index acceleration for jsonb_path_exists filter if it takes a variable list in its parameters. For example, `jsonb_path_exists(b, '$.a ? (@.b == $x)', '{"x": "c"}')` (where `'{"x": "c"}'` is the variable list) is not allowed.

Release note: None

ZhouXing19 committed:
sql/jsonpath: validate JSON column in jsonb_path_exists for inverted index acceleration

Informs: #154729

Previously, the jsonb_path_exists function would attempt inverted index
acceleration even when the JSON parameter referenced a column that was not
the source column of the inverted index, or when it was a constant JSON
value. This resulted in incorrect query plans and potentially wrong results.

This commit adds proper validation to ensure that inverted index acceleration
is only used when the JSON parameter is a variable expression that references
the actual source column of the inverted index. For all other cases, the
query will fall back to a full table scan with post-filtering.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154783)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice catch!

> @yuzefovich reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154783#-:-Oafbe-MAN0WbjJOYjTg:b-73m6vw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`-- commits` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154783#-OafaJUq6NNNRfowJ3nI:-OafaJUq6NNNRfowJ3nJ:b2yi46u):*
nit: no need for a release note given that we haven't released any binaries with the bug.
___
*[`pkg/sql/opt/invertedidx/json_array.go` line 434 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154783#-OafbaEvBcBSSyOHvF_t:-OafbaEvBcBSSyOHvF_u:bcrrj5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/af87c69038db42bf71878b1748a1cd78d19da6ea/pkg/sql/opt/invertedidx/json_array.go#L434)):*
> ```Go
> 			if len(t.Args) > 1 {
> 				// The first parameter has to be a column reference.
> 				if varExpr, ok := t.Args[0].(*memo.VariableExpr); ok {
> ```

nit: can we use `isIndexColumn` helper method instead? Seems like it does exactly what we need.


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql/jsonpath: disallow index acceleration for jsonb_path_exists with variables

We don't allow index acceleration for jsonb_path_exists filter if it takes a variable
list in its parameters. For example, `jsonb_path_exists(b, '$.a ? (@.b == $x)', '{"x": "c"}')`
(where `'{"x": "c"}'` is the variable list) is not allowed.

Release note: None

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154783)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/opt/invertedidx/json_array.go` line 434 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154783#-OafbaEvBcBSSyOHvF_t:-OafdVHN1_Bo1PNMcZPF:bkd5dle) ([raw file](https://github.com/cockroachdb/cockroach/blob/af87c69038db42bf71878b1748a1cd78d19da6ea/pkg/sql/opt/invertedidx/json_array.go#L434)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: can we use `isIndexColumn` helper method instead? Seems like it does exactly what we need.
</blockquote></details>

Ah yeah, thanks for the pointer! Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> @yuzefovich reviewed 2 of 2 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154783)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)



<!-- Sent from Reviewable.io -->
